opentelemetry-sdk/src/resource/builder.rs

```diff
@@ -0,0 +1,174 @@
+use std::borrow::Cow;
+
+use opentelemetry::{KeyValue, Value};
+
+use super::{Resource, ResourceDetector, SERVICE_NAME};
+
+/// Builder to allow easy composition of a Resource
+#[derive(Debug)]
+pub struct ResourceBuilder {
+    resource: Resource,
+}
+
+impl Default for ResourceBuilder {
+    /// Create ResourceBuilder with [Resource::default()].
+    ///
+    /// The default resource will contain the following detectors:
+    /// - SdkProvidedResourceDetector
+    /// - TelemetryResourceDetector
+    /// - EnvResourceDetector
+    fn default() -> Self {
+        ResourceBuilder {
+            resource: Resource::default(),
+        }
+    }
+}
+
+impl ResourceBuilder {
+    /// Create ResourceBuilder with [Resource::empty()].
+    pub fn new() -> Self {
```

Comment:
is this needed, given Resource already expose public methods to create a ResourceBuilder?

Reply:
wondering what is the best option?
Should users have the ability to get Builder instance from Builder itself, or via the thing it is ultimately building or both? We need to review rest of the codebase to ensure we do this consistently across.

I think its best to offer just one way to create a builder, and that must be via a method on the thing we are ultimately  building.
so Resource::builder().....build() - only this should be required.
and not ResourceBuilder::new()....build().

Reply:
It looks like the others are keeping new private, and exposing the builder through the api. In this case how would be want to handle starting with empty resource vs a default one? Would we want to always start with a empty one and then allow `with_default()` or something like that?

Reply:
the others are hiding the new method, I will update this, I agree with you.

Reply:
I think the reason that was public is to allow the creation of an empty resource vs a default resource. In this case would it make more sense to make it `pub(super)`? so the resource is still able to use it? Or only allow starting from an empty resource, and then expose a `with_default()`?